#8492 #8513 - Add a FuelCellElectricGenerator Unit Test and do not zero out Constant Skin Loss Rate in Generator:FuelCell:PowerModule

jmarrec opened the pull request:
Pull request overview
---------------------
 - Fixes #8492 - Add a FuelCellElectricGenerator Unit Test
 - Fixes #8513 - Generator:FuelCell:PowerModule: Constant Skin Loss Rate overriden to zero
 - Fixes #8514 - Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored

**Diffs are explained below**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Lint existing test, and add basic testing that it loads it fine (failing)

compare_err_stream:     Which is: "   ** Severe  ** <root>[Generator:FuelCell][Generator Fuel Cell 1] - Missing required property 'heat_exchanger_name'.\n"

jmarrec committed:
Ok, this appears to be a working version (except for one WaterHeater CalcStandardRatings warning)

jmarrec committed:
Write actual tests for components. Found one specific bug

jmarrec commented on tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc:
```diff
@@ -55,132 +55,1027 @@
+
+    EXPECT_EQ(0.6392, fCPM.RadiativeFract);
+
+    // TODO: this is an actual bug. It is being overriden in FigureFuelCellZoneGains, while here the SkinLossMode is ConstantRate, so it should be
+    // unaffected throughout
+    EXPECT_EQ(729, fCPM.QdotSkin);
```
TODO: this seems like a bug.

It's being overriden to 0.0 here: https://github.com/NREL/EnergyPlus/blob/f754bc5c2c5d415789c9b01297c575772833b862/src/EnergyPlus/FuelCellElectricGenerator.cc#L3343

jmarrec replied:
Addressed it in https://github.com/NREL/EnergyPlus/pull/8511/commits/4f837e59b1648cb73aeeb7af39dfb81ea96affcd

jmarrec commented on tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc:
```diff
@@ -55,132 +55,1027 @@
+    EXPECT_EQ(1.0, airSup.BlowerHeatLossFactor);
+    EXPECT_EQ(DataGenerators::AirSupRateMode::ConstantStoicsAirRat, airSup.AirSupRateMode);
+
+    // TODO: Interestingly, Stoics ratio is the input + 1.0
+    EXPECT_EQ(2.0, airSup.Stoics);
```
haven't really had time to track how that makes sense, but will do.

jmarrec replied:
Well, at least the IO ref mentions it, so no problem: https://bigladdersoftware.com/epx/docs/9-4/input-output-reference/group-electric-load-center-generator.html#field-stoichiometric-ratio

jmarrec commented on tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc:
```diff
@@ -55,132 +55,1027 @@
+        "  ConstantRate,                           !- Skin Loss Calculation Mode",
+        "  Thermal Zone 1,                         !- Zone Name",
+        "  0.6392,                                 !- Skin Loss Radiative Fraction",
+        "  729,                                    !- Constant Skin Loss Rate {W}",
```
FCMP.QdotSkin = 729

jmarrec committed:
Do not zero out FCPM.QdotSkin when Loss Mode = 'Constant Rate', and clean out test

jmarrec committed:
Correct test for NumCycles. It fails... another bug?

jmarrec committed:
Fix #8514 - Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored

jmarrec commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -266,7 +266,8 @@ namespace FuelCellElectricGenerator {
                 state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NomPel = NumArray(2);
-                state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCycles = NumArray(3);
+                state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCyclesAtStart = NumArray(3);
+                state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCycles = state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCyclesAtStart;
```
Fix #8514 - Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored

jmarrec commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -3194,7 +3195,7 @@ namespace FuelCellElectricGenerator {
             this->AirSup.QskinLoss = 0.0;
             this->AirSup.QintakeRecovery = 0.0;
-            this->FCPM.NumCycles = 0;
+            this->FCPM.NumCycles = this->FCPM.NumCyclesAtStart;
```
Fix #8514 - Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored

jmarrec commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -3340,7 +3341,10 @@ namespace FuelCellElectricGenerator {
                 cell.WaterSup.QskinLoss = 0.0;
                 cell.AuxilHeat.QskinLoss = 0.0;
-                cell.FCPM.QdotSkin = 0.0;
+                if (cell.FCPM.SkinLossMode != DataGenerators::SkinLoss::ConstantRate) {
+                    // If Constant Skin Loss Rate, then do not zero out
+                    cell.FCPM.QdotSkin = 0.0;
+                }
```
Fix #8513

jmarrec committed:
Register new advanced output var to test FCPM NumCycles changes and FCPM Skin Loss

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8492_FuellCell_Test

jmarrec commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -1330,6 +1331,20 @@ namespace FuelCellElectricGenerator {
                                 "Sum",
                                 this->Name);
+
+            SetupOutputVariable(state, "Generator Number of Cycles",
+                                OutputProcessor::Unit::None,
+                                this->Report.NumCycles,
+                                "System",
+                                "Average",
+                                this->Name);
+
+            SetupOutputVariable(state, "Generator Power Module Skin Heat Loss Rate",
+                                OutputProcessor::Unit::W,
+                                this->Report.FCPMSkinLoss,
+                                "System",
+                                "Average",
+                                this->Name);
```
Register two new **Advanced** report output variables to test changes for #8513 and #8514 

You need this to be able to see them:

```
  Output:Diagnostics,
    DisplayAdvancedReportVariables;
```

jmarrec commented:
# Explaining CI diffs in MicroCogeneration

To explain the **expected differences** that CI reports for MicroCogeneration, I added a modified `MicroCogeneration.idf` on EnergyPlusDevSupport:

https://github.com/NREL/EnergyPlusDevSupport/blob/d9b935df34d66e70773ab003a2ddbddd09e002ab/DefectFiles/8000s/8492/MicroCogeneration.idf#L5170-L5182

I also cherry picked https://github.com/NREL/EnergyPlus/pull/8511/commits/ccaad9edc27e197758b974290562c16f5d6de124 onto develop, so I would have the new output variables I registered available.

I ran an annual simulation with both with output variables added, and here are the results.


## #8514 - FCPM Number of cycles

I set the FCPM `Number of Stops at Start of Simulation` to **10**.

https://github.com/NREL/EnergyPlusDevSupport/blob/d9b935df34d66e70773ab003a2ddbddd09e002ab/DefectFiles/8000s/8492/MicroCogeneration.idf#L4859

As you can see, before it was zero, now it's properly 10.

![image](https://user-images.githubusercontent.com/5479063/107616657-cb5c3a00-6c4e-11eb-9b28-a6d7174618e6.png)

![image](https://user-images.githubusercontent.com/5479063/107617201-c3e96080-6c4f-11eb-98ad-0cf2cbf7f37b.png)


## #8513 FCPM Constant Skin Loss

The FCPM `Constant Skin Loss Rate {W}` is **729**: https://github.com/NREL/EnergyPlusDevSupport/blob/d9b935df34d66e70773ab003a2ddbddd09e002ab/DefectFiles/8000s/8492/MicroCogeneration.idf#L4878

Looking at the "Generator Zone Sensible Heat Transfer Rate", you can see that there is indeed 729 W accounted for now, that where missing before:

![image](https://user-images.githubusercontent.com/5479063/107617086-88e72d00-6c4f-11eb-9c09-912f013a38bf.png)


And using the specific advanced output variable I added, it's crystal clear:

![image](https://user-images.githubusercontent.com/5479063/107617160-a87e5580-6c4f-11eb-9cb5-983a994094c4.png)


![image](https://user-images.githubusercontent.com/5479063/107617176-b2a05400-6c4f-11eb-82df-da26aa267c34.png)

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8492_FuellCell_Test

Myoldmopar commented:
This looks really good to me.  I'll build it locally with all the changes to develop since the last commit and we'll see if this can go right in.  Thanks @jmarrec

dareumnam commented:
This looks really good to me too. I built it locally and ran the unit test successfully. I ran the example file and it runs cleanly, showed the result as @jmarrec mentioned in the comment above. I think the only thing that should be done before this gets merged in is adding the explanation about the two new output variables in I/O Reference Doc. Thank you @jmarrec

dareumnam commented on tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc:
```diff
@@ -55,132 +55,970 @@
+    EXPECT_TRUE(generatorController->electProdRate > 1.15 * generatorController->thermProdRate) << "Power to Heat Ratio appears too low";
+    EXPECT_DOUBLE_EQ(exhaustHX.qHX, generatorController->thermProdRate);
+    EXPECT_DOUBLE_EQ(generatorController->thermProdRate * 15 * 60, generatorController->thermalProd);
 }
```
Thank you so much for finishing this huge unit test and catching two more defects about this while at it.  We appreciate it.

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8492_FuellCell_Test

jmarrec committed:
Add advanced variables to IO/ref

jmarrec commented:
@dareumnam Thanks for the review and the catch about I/O ref. I added the variables in https://github.com/NREL/EnergyPlus/pull/8511/commits/2babab7296e646d8807ee440e245fce71424d8ba:

*  in the list of advanced report variables:

![image](https://user-images.githubusercontent.com/5479063/109633800-d406a880-7b48-11eb-9380-dd399919d571.png)

* And added a description with some links where appropriate and testing that the links do work correctly:

![image](https://user-images.githubusercontent.com/5479063/109633881-e7b20f00-7b48-11eb-9986-d1b215da0a21.png)

dareumnam commented:
Doc looks great, CI is all happy here. I'll merge this now. Thanks so much @jmarrec for the fix.
